src/Lean/Server/FileWorker.lean

```diff
@@ -470,6 +470,24 @@ section RequestHandling
     ⟨text.utf8PosToLspPos <| stx.getPos?.get!,
      text.utf8PosToLspPos <| stx.getTailPos?.get!⟩
 
+  open Elab in
+  partial def handlePlainTermGoal (p : PlainTermGoalParams)
+    : ServerM (Task (Except IO.Error (Except RequestError (Option PlainTermGoal)))) := do
+    let st ← read
+    let doc ← st.docRef.get
+    let text := doc.meta.text
+    let hoverPos := text.lspPosToUtf8Pos p.position
+    withWaitFindSnap doc (fun s => s.endPos > hoverPos)
+      (notFoundX := pure none) fun snap => do
+        for t in snap.cmdState.infoState.trees do
+          if let some (ci, Info.ofTermInfo i) := t.termGoalAt? hoverPos then
+            let goal ← ci.runMetaM i.lctx <| open Meta in do
+              let ty ← instantiateMVars <| i.expectedType?.getD (← inferType i.expr)
```

Comment:
I wonder whether we should fall back on the inferred type, or return nothing if the expected type isn't present, but try to make sure it's always present. I'm assuming there can be cases where the expected and inferred type are defeq but not the same. Which variant would be clearer to users then?

Reply:
The code above looks reasonable to me. @Vtec234 If we find examples where it is producing weird results, please create an issue.